src/lib.rs

```diff
@@ -80,24 +98,79 @@ pub mod __rt {
     /// the futures protocol.
     struct GenFuture<T>(T);
 
+    /// Small shim to translate from a generator to a stream.
+    struct GenStream<T> {
+        gen: T,
+        done: bool,
+    }
+
+    /// Uninhabited type to allow `await!` to work across both `async` and
+    /// `async_stream`.
+    pub enum Mu {}
+
     pub fn gen<T>(gen: T) -> impl MyFuture<T::Return>
-        where T: Generator<Yield = ()>,
+        where T: Generator<Yield = Async<Mu>>,
               T::Return: IsResult,
     {
         GenFuture(gen)
     }
 
+    pub fn gen_stream<T, U>(gen: T) -> impl MyStream<U, T::Return>
+        where T: Generator,
+              T::Yield: IsAsync,
+              <T::Yield as IsAsync>::Item: IsResult<Ok = U>,
+              T::Return: IsResult<
+                  Ok = (),
+                  Err = <<T::Yield as IsAsync>::Item as IsResult>::Err
+              >,
+    {
+        GenStream { gen, done: false }
+    }
+
     impl<T> Future for GenFuture<T>
-        where T: Generator<Yield = ()>,
+        where T: Generator<Yield = Async<Mu>>,
               T::Return: IsResult,
     {
         type Item = <T::Return as IsResult>::Ok;
         type Error = <T::Return as IsResult>::Err;
 
         fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
             match self.0.resume() {
-                GeneratorState::Yielded(()) => Ok(Async::NotReady),
-                GeneratorState::Complete(e) => e.into_result().map(Async::Ready),
+                GeneratorState::Yielded(Async::NotReady)
+                    => Ok(Async::NotReady),
+                GeneratorState::Yielded(Async::Ready(_))
+                    => unreachable!("Mu doesn't exist"),
+                GeneratorState::Complete(e)
+                    => e.into_result().map(Async::Ready),
+            }
+        }
+    }
+
+    impl<T> Stream for GenStream<T>
+        where T: Generator,
+              T::Yield: IsAsync,
```

Comment:
Is the `IsAsync` trait needed here? Or can this just use `T: Generator<Yield = Async<U>>`?

Reply:
I believe it's required to allow yielding results and have the equality constraint `<<T::Yield as IsAsync>::Item as IsResult>::Err = <T::Return as IsResult>::Err`. If there's some way to encode that equality without it then it can be dropped, I'll take a look at all these bounds and see if there's some sort of simplification that can happen.

Reply:
Simplified it to

```rust
where
    T: Generator,
    T::Yield: IsAsync<Item = Result<U, <T::Return as IsResult>::Err>>,
    T::Return: IsResult<Ok = ()>,
```

I attempted

```rust
where
    T: Generator<Yield = Async<Result<U, <T::Return as IsResult>::Err>>>,
    T::Return: IsResult<Ok = ()>,
```

but that fails because of `error[E0391]: unsupported cyclic reference between types/traits detected`